compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -2965,6 +2979,190 @@ static void generateArrayElementLoad(TR::Node* node, TR::Register* valueReg, uin
    generateRegMemInstruction(loadOpCode,  node, valueReg, generateX86MemoryReference(addressReg, index, cg), cg);
    }
 
+/** \brief
+*    Generate instructions to do array copy with copy size as a constant
+*
+*  \param node
+*     The tree node
+*
+*  \param dstReg
+*     The destination array address register
+*
+*  \param srcReg
+*     The source array address register
+*
+*  \param sizeReg
+*     The register holding the total size of elements to be copied, in bytes
+*
+*  \param elementSize
+*     The size of an element, in bytes
+*
+*  \param copySize
+*     The copy size, in bytes
+*
+*  \param cg
+*     The code generator
+*/
+static void arrayCopyPrimitiveInlineSmallSizeConstantCopySize(TR::Node* node,
+                                                              TR::Register* dstReg,
+                                                              TR::Register* srcReg,
+                                                              TR::Register* sizeReg,
+                                                              uint32_t copySize,
+                                                              TR::CodeGenerator* cg)
+   {
+   if (cg->comp()->getOption(TR_TraceCG))
+      {
+      traceMsg(cg->comp(), "%s: node n%dn srcReg %s dstReg %s sizeReg %s copySize %d\n", __FUNCTION__, node->getGlobalIndex(),
+         cg->comp()->getDebug()->getName(srcReg), cg->comp()->getDebug()->getName(dstReg), cg->comp()->getDebug()->getName(sizeReg), copySize);
+      }
+
+   if (copySize == 0)
+      return;
+
+   TR::Register* tmpReg1 = cg->allocateRegister(TR_GPR);
+   TR::Register* tmpReg2 = cg->allocateRegister(TR_GPR);
+   TR::Register* tmpVRFReg1 = cg->allocateRegister(TR_VRF);
+   TR::Register* tmpVRFReg2 = cg->allocateRegister(TR_VRF);
+
+   if ((copySize == 1) || (copySize == 2) || (copySize == 4) || (copySize == 8) || (copySize == 16) || (copySize == 32)
+       || ((copySize == 64) && cg->comp()->target().cpu.supportsFeature(OMR_FEATURE_X86_AVX512F)))
+      {
+      TR::Register* valueReg = (copySize >= 16) ? tmpVRFReg1 : tmpReg1;
+      generateArrayElementLoad(node, valueReg, copySize, srcReg, 0 /* index */, cg);
+      generateArrayElementStore(node, dstReg, 0 /* index */, valueReg, copySize, cg);
+      }
+   else
+      {
+      uint8_t regSize = 0;
+      TR::Register* t1 = tmpReg1;
+      TR::Register* t2 = tmpReg2;
+
+      if (copySize < 8)
+         {
+         regSize = (copySize < 4) ? 2 : 4;
+         }
+      else if (copySize < 16) // 8-15
```

Comment:
Is it true that `copySize` is always the full length of the destination array?  That is, `copySize` is not a subsection of the array.  If the answer is no, then I don't think this copy works in all cases because you may overwrite more bytes in the destination than you were allowed to copy.  For example, let's say `copySize` is 13.  This will generate two 8-byte loads and stores from source to destination.  But I think that only works if you know that `13` is also the full length of the destination array (where you can copy some extra bytes into the slush between objects).

Reply:
> Is it true that copySize is always the full length of the destination array? 

No, the copy size isn't always the full length of the detonation array. It doesn't need to be 

If `copySize` is 13, the generated load/store instructions from `generateMemoryCopyInstructions` as below: 
```
 [0x7f485307c070] mov   rdx, qword ptr [rax+1*rdi-0x8]      # L8RegMem, SymRef [#4758 -8]
 [0x7f485307c190] mov   rax, qword ptr [rax]    # L8RegMem
 [0x7f485307c2b0] mov   qword ptr [rcx+1*rdi-0x8], rdx      # S8MemReg, SymRef [#4760 -8]
 [0x7f485307c3d0] mov   qword ptr [rcx], rax    # S8MemReg
```

```
src:  |------- copysize=13 ------|
..... 1 2 3 4 5 6 7 8 9 10 11 12 13 .....
              |------------------|       
                first load 8 bytes at src + (13 - 8) = src + 5
      |--------------|
         second load 8 bytes at src

dst:  |------- copysize=13 ------|
..... 1 2 3 4 5 6 7 8 9 10 11 12 13 .....
              |------------------|       
                first store 8 bytes at dst + (13 - 8) = dst + 5
      |--------------|
         second store 8 bytes at dst 
```